src/coreclr/src/jit/emit.cpp

```diff
@@ -5701,17 +5742,21 @@ void emitter::emitDataGenEnd()
  *
  * Returns constant number as offset into data section.
  */
-UNATIVE_OFFSET emitter::emitDataConst(const void* cnsAddr, UNATIVE_OFFSET cnsSize, UNATIVE_OFFSET cnsAlign)
+UNATIVE_OFFSET emitter::emitDataConst(const void* cnsAddr, unsigned cnsSize, unsigned cnsAlign, var_types dataType)
 {
-    UNATIVE_OFFSET cnum = emitDataGenBeg(cnsSize, cnsAlign);
-    emitDataGenData(0, cnsAddr, cnsSize);
-    emitDataGenEnd();
+    UNATIVE_OFFSET cnum = emitDataGenFind(cnsAddr, cnsSize, cnsAlign, dataType);
```

Comment:
You've introduced an O(n) search for each constant. Presumably, you could construct a function with a large number of constants where this might show up on a trace. It seems like the usual case would be very few searches, but the usual case is also probably very few matches (duplicates). Would be interesting to see some stats about this across our SPMI collections.

Reply:
Yes, I understand that.  There are very few constants used by the JIT, so I didn't think it was worth adding a hash table for this.  I can add a check that counts the total number and asserts (in a checked JIT) when it exceeds a certain limit, (like 10,000 or so)

Reply:
I'm not sure it's worth an assert. I suppose the search function could bail out after, say, 100 constants and just always add a new (possibly duplicated) constant after that.

Reply:
That is another possibility.  Let me run the tests and SPMI to see if we even have this case today.
Using 10k we would hits this after have 142 unique constants and/or switch tables in a single method.

Reply:
I settled on only checking for matches in the first 64 entries and the adding the new value if is doesn't match

```
// If we don't find a match in the first 64, then we just add the new constant
// This prevents an O(n^2) search cost
```